common/src/main/java/com/wynntils/mc/mixin/ClientPacketListenerMixin.java

```diff
@@ -49,17 +71,85 @@ public abstract class ClientPacketListenerMixin {
+
+    @Shadow
+    @Final
+    private Minecraft minecraft;
+
+    @Shadow
+    public abstract FeatureFlagSet enabledFeatures();
+
+    @Shadow
+    private LayeredRegistryAccess<ClientRegistryLayer> registryAccess;
+
+    @Shadow
+    private MessageSignatureCache messageSignatureCache;
+
+    @Shadow
+    private SignedMessageChain.Encoder signedMessageEncoder;
+
+    @Shadow
+    private LastSeenMessagesTracker lastSeenMessages;
+
     private static boolean isRenderThread() {
-        return (McUtils.mc().isSameThread());
+        return McUtils.mc().isSameThread();
+    }
+
+    @Inject(method = "sendChat", at = @At("HEAD"), cancellable = true)
+    private void onChatPre(String string, CallbackInfo ci) {
+        ChatSentEvent result = EventFactory.onChatSent(string);
+        if (result.isCanceled()) {
+            ci.cancel();
+        }
+
+        if (!Objects.equals(string, result.getMessage())) {
+            // Note: This code is taken from the MC source code. I don't think there is a sane way of not doing this,
+            // unfortunately (in 1.19.2, there was, not in 1.19.3).
+            //       For anyone concerned, this does not mess with NoChatReport at the time of writing.
+
+            Instant instant = Instant.now();
+            long l = Crypt.SaltSupplier.getLong();
+            LastSeenMessagesTracker.Update update = this.lastSeenMessages.generateAndApplyUpdate();
+            MessageSignature messageSignature =
+                    this.signedMessageEncoder.pack(new SignedMessageBody(string, instant, l, update.lastSeen()));
+            this.send(new ServerboundChatPacket(string, instant, l, messageSignature, update.update()));
+
+            ci.cancel();
+        }
+    }
+
+    @Inject(method = "sendCommand", at = @At("HEAD"), cancellable = true)
+    private void onSignedCommandPre(String string, CallbackInfo ci) {
+        if (EventFactory.onCommandSent(string, true).isCanceled()) {
+            ci.cancel();
+        }
+    }
+
+    @Inject(method = "sendUnsignedCommand", at = @At("HEAD"), cancellable = true)
+    private void onUnsignedCommandPre(String command, CallbackInfoReturnable<Boolean> cir) {
+        if (EventFactory.onCommandSent(command, false).isCanceled()) {
+            cir.setReturnValue(false);
+            cir.cancel();
+        }
     }
 
     @Inject(
             method = "handleCommands(Lnet/minecraft/network/protocol/game/ClientboundCommandsPacket;)V",
             at = @At("HEAD"))
     private void handleCommandsPre(ClientboundCommandsPacket packet, CallbackInfo ci) {
         if (!isRenderThread()) return;
-        CommandsPacketEvent event = EventFactory.onCommandsPacket(packet.getRoot());
-        ((ClientboundCommandsPacketAccessor) packet).setRoot(event.getRoot());
+        RootCommandNode<SharedSuggestionProvider> root = packet.getRoot(
+                CommandBuildContext.simple(this.registryAccess.compositeAccess(), this.enabledFeatures()));
```

Comment:
This looks just like a magic incantation. Did you find this on the internet, or figure it out yourself? If the latter, how? :-O Or rather, what does this do? If you've just copied it from somewhere, I'm happy to accept that too :)

Reply:
I got this from either reading MC source code, or some FAPI source code.